lib/src/client.dart

```diff
@@ -2429,50 +2435,66 @@ class Client extends MatrixApi {
-          final String? redacts = eventUpdate.content.tryGet<String>('redacts');
-          if (redacts != null) {
-            room.states.forEach(
-              (String key, Map<String, Event> states) => states.forEach(
-                (String key, Event state) {
-                  if (state.eventId == redacts) {
-                    state.setRedactionEvent(stateEvent);
-                  }
-                },
-              ),
-            );
-          }
-        } else {
-          // We want to set state the in-memory cache for the room with the new event.
-          // To do this, we have to respect to not save edits, unless they edit the
-          // current last event.
-          // Additionally, we only store the event in-memory if the room has either been
-          // post-loaded or the event is animportant state event.
-          final noMessageOrNoEdit = stateEvent.type != EventTypes.Message ||
-              stateEvent.relationshipType != RelationshipTypes.edit;
-          final editingLastEvent =
-              stateEvent.relationshipEventId == room.lastEvent?.eventId;
-          final consecutiveEdit =
-              room.lastEvent?.relationshipType == RelationshipTypes.edit &&
-                  stateEvent.relationshipEventId ==
-                      room.lastEvent?.relationshipEventId;
-          final importantOrRoomLoaded =
-              eventUpdate.type == EventUpdateType.inviteState ||
-                  !room.partial ||
-                  // make sure we do overwrite events we have already loaded.
-                  room.states[stateEvent.type]
-                          ?.containsKey(stateEvent.stateKey ?? '') ==
-                      true ||
-                  importantStateEvents.contains(stateEvent.type);
-          if ((noMessageOrNoEdit || editingLastEvent || consecutiveEdit) &&
-              importantOrRoomLoaded) {
-            room.setState(stateEvent);
+        room.setState(Event.fromJson(eventUpdate.content, room));
+        break;
+      case EventUpdateType.state:
+        final event = Event.fromJson(eventUpdate.content, room);
+        if (event.stateKey != null) break;
+        // For invite states or post loaded rooms, we set everything in-memory.
+        // For other rooms we only set "important" state events.
+        if (!room.partial ||
+            // make sure we do overwrite events we have already loaded.
+            room.states[event.type]?.containsKey(event.stateKey ?? '') ==
+                true ||
+            importantStateEvents.contains(event.type)) {
+          room.setState(event);
+        }
+        break;
+      case EventUpdateType.timeline:
+        final event = Event.fromJson(eventUpdate.content, room);
+        // For invite states or post loaded rooms, we set everything in-memory.
+        // For other rooms we only set "important" state events.
+        if (event.stateKey != null &&
+            (!room.partial || importantStateEvents.contains(event.type))) {
```

Comment:
You seem to be duplicating a lot of code between the "state" and "timeline" cases. That makes it rather hard to spot bugs. For example this would not update state events we already store in memory.

Reply:
Fixed